pkg/datapath/loader/netlink.go

```diff
@@ -170,7 +170,11 @@ func attachProgram(link netlink.Link, prog *ebpf.Program, progName string, qdisc
 		// Omitting XDP_FLAGS_UPDATE_IF_NOEXIST equals running 'ip' with -force,
 		// and will clobber any existing XDP attachment to the interface.
 		if err := netlink.LinkSetXdpFdWithFlags(link, prog.FD(), int(xdpFlags)); err != nil {
-			return fmt.Errorf("attaching XDP program to interface %s: %w", link.Attrs().Name, err)
+			if option.Config.NodePortAcceleration == option.XDPModeBestEffort {
```

Comment:
I don't like swallowing the error here; this function is not specific to NodePorts. Rather, this config check should be specific to the code path that attaches the NodePort XDP program

Reply:
While not entirely specific to NodePort code path it could live in `reinitializeXDPLocked()`. I think the original PR #25870 placed its logic there as well. Plus there is already logic to skip certain devices in that function.

Reply:
Yeah, this section felt like a pretty good fit when I was going through everything. Happy to adjust if there are strong opinions about it.

Reply:
I'll let @cilium/loader make the final decision.

Reply:
A good rule of thumb to apply is to avoid use of `option.Config` at deeper layers since that makes testing fairly annoying since now in tests you'll need to mutate the global `option.Config` and then restore it. 

One possibility here is to just pass the `option.Config.XDPMode` that `reinitializeXDPLocked` grabs all the way down to `attachProgram` and handle the "BestEffort" option there and then convert to XDP flags. This way this low-level function isn't coupled to the `option.Config` global. E.g. just move the "xdpModeToFlags" from `replaceDatapath` to `attachProgram`.

Also talking about tests, could you look into adding a test case for this into `netlink_test.go`? `TestAttachProgram` already has a "XDP" sub-test which would be easy to copy + adapt.

Reply:
Also I wonder if skipping is the right thing to do here. We can also try the generic mode before failing and we can also fall back to TC.

Reply:
@joamaki From the documentation it doesn't sound like Generic is recommended outside of testing scenarios.

Taken from https://docs.cilium.io/en/stable/bpf/progtypes/

```
For drivers not implementing native or offloaded XDP yet, the kernel provides an option for generic XDP which does not require any driver changes since run at a much later point out of the networking stack. This setting is primarily targeted at developers who want to write and test programs against the kernel’s XDP API, and will not operate at the performance rate of the native or offloaded modes. For XDP usage in a production environment either the native or offloaded mode is better suited and the recommended way to run XDP.
```

As this PR is currently written it will fall back to tc if XDP "native" fails.

Reply:
Does anyone else want to weigh in on this?

Reply:
> Also I wonder if skipping is the right thing to do here. We can also try the generic mode before skipping entirely. OTOH, quite likely we'd prefer the tc filter program as a fallback to generic XDP... Perhaps someone from sig-loader can chime in on whether it'd make sense or not.

Generic mode should not be used in production, it will bypass GRO engine. Fallback to tc is significantly better.

Reply:
> Generic mode should not be used in production, it will bypass GRO engine. Fallback to tc is significantly better.

Agreed. Also, adding a retry loop will make the pending changes in #28308 a bit more painful to integrate. Making best-effort mode result in 'optional' XDP progs will result in the least friction overall.

#28308 lifts the `LinkSetXdpFdWithFlags()` here into a new `attachXDPProgram()` function and renames `attachProgram()` to `attachTCProgram()`, so the if statement here will have to move to the call site in `replaceDatapath()`. That all looks :+1:.